### DISCUSSION TITLE
[Feature Request] Launcher GUI

### DISCUSSION THREAD

**Discussion by Darthagnon** [MikeMaximus/gbm#209] [Ideas]
General GUI suggestion/feature request:

Would it be possible to add more to the launcher GUI, possibly something like this?

![SmartSteamEmu Comfy Launcher](https://user-images.githubusercontent.com/20792748/101659461-fc956f80-3a3d-11eb-8edb-37384ad23267.PNG) 

([Source](https://cs.rin.ru/forum/viewtopic.php?f=10&t=91948), click on the "Old version" spoiler dropdown)

I use the context-menu launcher on the taskbar icon, but I don't find the Ctrl+L launcher search bar super useful at the moment because it's not a global shortctut, and not useful for browsing a list of games.

GBM already has the tech built-in to display extracted game EXE icons, names, paths and playtime data (as in the current GBM main GUI showing "last played game"). Was wondering if maybe the current main GUI could stay more-or-less the same, but have a scrollable sidebar with the gamelist added (as above), with the Ctrl+L search bar at the top of the GUI to filter the gamelist. By default it could show last-played or "no game detected" screen, but clicking on a name in the sidebar could populate the main view with the info that is currently displayed once you finish playing a game. Add a "Play" launch button  to the main view, to (re)launch the current game, and you've got a full launcher to rival Steam and Playnite.

**Comment by MikeMaximus** [comment]
I do like this idea.

The GUI features for the launcher are pretty rough as I was concentrating more on functionality for this initial release.  I was also wary about making large GUI changes to support it, as I wasn't sure how it was going to be received.  I always get annoyed when I'm using an app for one purpose, then the developers add new features and try to force them on me when I don't want them.  This is why I always make new GBM features not directly related to saved game backups disabled and hidden by default.

I'll think about it and see what I can come up with.  Thanks for the suggestion!

**Comment by Darthagnon** [comment]
Glad you like it.

I really do appreciate what you say about developers messing around with features and GUI and stuff; I still haven't forgotten how beautiful Chromium used to be, then Google took a steaming dump on it and called it "Material™", or what Valve has done lately with Steam.

Personally, though, in this case, I feel it would be an improvement with no downsides, as one thing I always struggled about with other launchers (e.g. PlayNite) is "Why would I import my games, just to have them in a UI?" With GBM, I import all my games so as to have backups of their saves, so a simple, playable UI to display the (already imported) games would be big icing on the cake! 

Here's a quick mockup I did in Paint.NET of what I imagined:
![GBM Launcher mockup](https://user-images.githubusercontent.com/20792748/101674228-248dce80-3a50-11eb-9bea-287003784039.png)


A key advantage of this over just using the "Game setup" screen to launch games is it isn't editable and doesn't have lots of buttons. Once setup, just one or two clicks to launch any imported game.

**Reply by MikeMaximus** [reply]
The Game Manager has become so busy, even without the launcher it would be logical to do this and move some of the most used features directly onto the main window.  I could make the side bar collapsible as well so it can still be a minimal interface.

**Reply by Darthagnon** [reply]
After you mentioned the busyness of the Game Manager, I had a think, and drew up another mockup for possible simplification (simple, and expanded with comments versions attached). EDIT: I merged the ```Monitor Only``` and ```Monitor this game``` buttons into ```Automatically backup```, as that was roughly what I understood from the documentation. I haven't really used these checkboxes, though, as I want all games that I add to GBM to be monitored and backed up (the reason I use GBM), and I never really understood them.

![GBM mockup 2 simplification SIMPLE](https://user-images.githubusercontent.com/20792748/102020127-0fc27b00-3d6f-11eb-84b6-754277f204f8.png)

![GBM mockup 2 simplification annotated](https://user-images.githubusercontent.com/20792748/102020130-15b85c00-3d6f-11eb-8741-46507840c88b.png)

**Reply by MikeMaximus** [reply]
Thanks for the suggestions!

- The "OS" field is mainly for Linux users because they can use Windows and Linux native configurations.  Using tags for that would be too unreliable.  Maybe I could hide this in Windows, but it would still need to be placed somewhere.

- "Monitor this Game" allows you to setup a backup configuration, but prevent it from automatically detected.  This was added so you could have some configurations that were manual backup only.  It's suppose it could be removed now that configurations can be setup with an empty process field, which accomplishes the same thing.

- "Monitor Only" configurations allow GBM to track the time and sessions of games that don't have anything to backup.  Such as MMOs or CCGs.

- I actually did the icon suggestion this morning in commit f56288b.  :smile: Icons will now be cached and displayed where appropriate after a game has been detected for at least one session.

- The "Backup Information" section has always been the mess I keep putting off fixing.  Clickable text links and smaller buttons is probably the way to go.  I also like the checkmark to replace Mark as Restored.

**Reply by Darthagnon** [reply]
Thank you so much for the explanations! I'll add those to the offline manual when I next update it (probably for the next release of GBM). I noticed the icon commit, and I was thinking "Wow, that's fast!" 😄 guess we were thinking on the same wavelength. I can try forking and taking a look at the source later this week, see if I can figure out anything to help (uncertain, as I've only tinkered with C# in the past).

- ✅ = White Heavy Checkmark Emoji, or, how my PC renders Unicode checkmark U+2705
- ❌= Cross Mark Emoji, or, how my PC renders Unicode U+274C
They might render differently on other systems, e.g. type it into Notepad and it's a B&W Win7 emoji; been tinkering with [SymWin](https://github.com/Darthagnon/SymWinBinaries) C# source code, and I've found some older emojis, like these, are accepted by certain types of code, whereas a lot of newer emojis are flat-out rejected

**Comment by MikeMaximus** [comment]
I started tinkering yesterday and ended up doing an entire pass of a new GUI on the main form.

![image](https://user-images.githubusercontent.com/15719732/101984386-ca0a9300-3c46-11eb-9f3d-512df8e6f967.png)

How's it looking?  The left and right panels can be resized, they will resize properly with the form, and the game panel can be collapsed.  In general it behaves exactly as it did before, except you can now directly access some of the most commonly used features without accessing the Game Manager.

It's currently functional on the v125 branch, but it'll be very buggy until I push another pass later today.

**Reply by Darthagnon** [reply]
That is looking beautiful! Thank you so much! ❤ Thank you for making a dream come true!

Three things I thought I'd mention:
1. Would it be possible for the Play button glyph to be green? 
2. **(not sure about this one)** And/or possibly separate from the other buttons, e.g. Top Right, after Game Title? or maybe it's just the right-alignment of the buttons, and the left alignment of everything else that feels a little odd. 
3. Edit glyph - pencil on notepad rather than just notepad? Suggests editing more.

**Reply by MikeMaximus** [reply]
I'm just using icons from a nice royalty free pack, those are the best matches I could find.  I agree about the button placements, it doesn't look good.  I've been playing with different layouts, I'll post another screenshot when I decide on something final.

**Reply by Darthagnon** [reply]
Not sure if GitHub compresses uploads (otherwise, I'll try a pull request later this week). Did a quick edit (hue + contrast adjustment for Play button, pencil from Icons8 resized really tiny and overlayed on the Edit notepad) - thought they might look a bit nicer like this? Restore = Red, as it is potentially destructive

![play](https://user-images.githubusercontent.com/20792748/102029007-b6753e80-3da4-11eb-9fb3-c72469fd1f2c.png)
![edit](https://user-images.githubusercontent.com/20792748/102029008-b6753e80-3da4-11eb-8924-d76e0d50b323.png)
![restore](https://user-images.githubusercontent.com/20792748/102029911-5ed8d200-3da8-11eb-908b-4efda48ab6d3.png)

**Reply by MikeMaximus** [reply]
It looks like the icons retained their quality.  They look great, thank you.

Here is what it's currently looking like.

![image](https://user-images.githubusercontent.com/15719732/102098100-615d1b00-3dec-11eb-95b5-661f57bee179.png)

**Comment by Darthagnon** [comment]
Loosely related question: Does the search function index tags? Or can it be made to?

**Reply by MikeMaximus** [reply]
The game manager and main window currently do not support tags in the search/quick filter, but they really should.  I'll do that.

**Reply by Darthagnon** [reply]
Thank you so much!

**Reply by MikeMaximus** [reply]
Tags are now included in the quick filter/search.  Unfortunately this has caused various performance hits due to tags needed to be loaded completely, instead of on demand.  I'll have to see if it can be optimized by reducing the need to completely reload lists.

**Reply by MikeMaximus** [reply]
I've rewritten the way quick filters are applied, instead of changing how tag data is loaded.  So now you can search by tag and it actually seems to perform a little better.  😮

**Comment by Darthagnon** [comment]
Because GitHub nests posts and makes them difficult to see, here's a new one.

### Game Manager simplification
I tried to implement something like the layout I prototyped above. Please find attached the edited files and Windows Emoji PNG files (from Emojipedia). I didn't do a pull request, as I was unable to get it to build on my PC (and hence, it is untested).

What I did:
- Attempted a workaround such that the Backup File and Restore Path fields should highlight text in blue and underline it, with hand cursor on hover with a tooltip, but the whole boxes are "buttons" that borrow the Open Path functions of the current buttons.
- Removed the "Monitor this game" checkbox, since I understood it was deprecated. Tried (probably unsuccessfully) to add hint text to the Process Field "Leave blank to disable auto monitoring"
- Put small buttons with ```Delete backup``` (❌), ```Mark as Restored``` (✅) and ```Import Save``` (📁folder, should use the blue folder icon already in GBM??) next to the fields in Backup Information. I wanted to add the coloured glyphs, but couldn't figure out how, so I left it with the B&W Unicode glyphs, and have attached the coloured versions as 24px PNGs. I was thinking it would be better merging the "Delete Backup" and "Import Save" buttons into one with a dropdown (e.g. like ```Link```), but the subroutine for the ```Delete``` button was more complex than the others, so I decided I'd better not mess with it.
- Renamed "Backup File:" to ```Backup:``` button, and "Restore Path:" to ```Restore:```, and made them into "Restore" and "Backup" buttons. Renamed so that their titles can mean both "This next field relates to Backup/Restore", but simultaneously mean "Click Here to Backup/Restore". Not 100% sure this is a good change though, without testing.
- Repositioned ```Launch Options```, ```Save``` and ```Cancel```

Here are the files (probably don't need ```frmGameManager.resx```, but it was reported as changed by Git, so I included it just in case)
[Changed Files.zip](https://github.com/MikeMaximus/gbm/files/5696437/Changed.Files.zip)

**Reply by MikeMaximus** [reply]
Thanks again for the suggestions.  I'm glad your enthusiastic to help out, but mockups help me out much more than directly changing any code.  It's much easier to look at a mockup and decide what to do, than try and import code changes and view that in the designer.

I've done one pass of the Game Manager changes.  This is how it currently looks.

![image](https://user-images.githubusercontent.com/15719732/102261924-fe48b280-3ed7-11eb-9959-4c1deabac822.png)

I'll go over what was changed or not changed.

1.  I'm still on the fence about removing "Monitor this Game", it needs more consideration because it involves changing the backend and cleaning up other code.
2.  The Tags button has been removed.  You now manage tags by clicking on them directly, if no tags exist the link shows "Manage Tags...".
3.  The "Links" button has been renamed to "Advanced" and contains options for "Link Process", "Link Configuration", and "Launch Settings".
4.  Button (Recycling Bin) beside Backup Data, contains options "Delete Selected", "Delete All", "Import Backup".
5.  Button (Checkmark) beside Local Data, replaces the Mark as Restored button.  I used your nice green icon,  but it's only shows Green when it can be clicked.
6.  Backup File is now a link, opens the backup file in the associated app.
7.  Button (Open Folder) beside Backup File, this opens the folder containing the backup file in explorer.
8.  Restore Path is now a link, opens the restore path in the OS file explorer.
9.  All the new buttons and links have tool tips describing what function they perform.  
10.  The links are black because I found blue just looked strange in relation to the rest of the controls.

I'm not sure yet what to do with the extra space, so I just left it for now.  I'm considering reorganizing the "Configuration" section for the next pass, in addition to other changes.

**Reply by Darthagnon** [reply]
Thank you so much for taking the suggestions (and for ignoring the bad ones; I wasn't sure about some of the edit I tried)! Glad to have been able to help in a tiny way. Point taken RE: code editing; I'm a better artist than programmer any day. 

It looks "practically perfect in every way". Nothing wrong with leaving a bit of empty space 😄 - in my opinion, reason GBM Manager was looking a bit busy beforehand was due to all the space being completely used, with multiple layers of buttons at the bottom. Looks cleaner and more logical now, all buttons and fields have obvious purposes, and don't have overwhelming stacks.

**Reply by MikeMaximus** [reply]
I've made a second pass.  I split the sections into tabs and moved some controls to where I think they make more sense.  I'm not sure if I went too far,  but I think it looks pretty clean.  I'll try out this new video pasting feature instead of multiple screenshots.

https://user-images.githubusercontent.com/15719732/102641777-40653480-4122-11eb-8ef5-a3a7347566af.mp4

**Reply by Darthagnon** [reply]
Personally, I prefer how it was in [the previous screenshot](https://user-images.githubusercontent.com/15719732/102261924-fe48b280-3ed7-11eb-9959-4c1deabac822.png). Tabs works and is logical, but I preferred having all the settings needed on one page. EDIT: Though I guess if tabs were kept, the "Advanced" button could be expanded into another tab or a subsection without causing problems. EDIT 2: Tabs are also more future-proof (possibly an inevitable necessity?) if many more features are planned or could be added.

**Reply by MikeMaximus** [reply]
I'm still playing with it, but it looks a bit cleaner now.  Like you said it's a bit more annoying to work with when data is split across tabs.

![image](https://user-images.githubusercontent.com/15719732/102813439-6abb2a00-438e-11eb-8755-b1b8ee35e01b.png)

**Reply by Darthagnon** [reply]
Wondering, is there a possibility of a toggle in Settings to "Expand tabs into sections"? e.g. if we have a large enough monitor and want to take advantage of vertical space, enable, and the tabs expand to become sections (similar to how it used to be)? Or, wondering, whether this could be automatic on resizing the Game Manager window vertically?

**Reply by MikeMaximus** [reply]
It's a good idea, but this would be quite difficult.  Especially making it work in Linux/Mono as well, which is extremely wonky when making form design changes at run time.  I have recently moved more fields into the core configuration section to reduce the amount of tab switching required when adding or editing configurations.

![image](https://user-images.githubusercontent.com/15719732/104847055-7e7d8180-58a3-11eb-9122-1cee7fe01602.png)

**Reply by AdriandeLima** [reply]
[Darthagnon here, using his brother's account to post this]:

I did another build of the latest version, and noticed, as per your screenshot, that the Game Manager was looking very blue-themed. I thought I'd try and colour some of the icons; still not 100% sure I've got the brightness/contrast right, but I think they look quite nice:

![GBM_JDGnqRtKUW](https://user-images.githubusercontent.com/68482408/105413479-11187a80-5c2e-11eb-8955-cda87b33aeaf.png)

Here's the modified icons with some rough notes in a TXT:
[GBM_colors.zip](https://github.com/MikeMaximus/gbm/files/5852200/GBM_colors.zip)

A couple of thoughts I had:
- Would text labels on the ```Monitor``` and ```Link``` buttons be possible, alongside the icons? I found I had to keep checking the tooltips as I couldn't remember what they were for. Will make references to them in the documentation easier.
- The ```Game ID``` button is supposed to be a Database platter stack, unless I'm mistaken, but at low resolutions, just looks like a blue rectangle; I'll try editing to look more identifiable, but might be easier just to label the button "ID"?

### Some recolours I was suggested, but unsure about
My little brother suggested I should recolor ```Export``` icon as purple, because it is a sort of "save as" function, and Save icons are usually purple. He also suggested I recolor ```Cancel``` as red, too.

**Reply by MikeMaximus** [reply]
I don't mind the blue theme, but varying the colors does make them stand out more.

When I put an adequate label on the Monitor and Link buttons, it required them to be quite large, which also looked strange.  I felt the tooltips worked well enough on all the unlabeled buttons, maybe I'll play around with those again.

Yes the Game ID is supposed to be a database platter, but it doesn't scale down well I agree.  I just couldn't find anything else adequate, and that was another case of the text just looking strange on that size of a button.

I like the suggestion of the red cancel button,  but leaving the import and export icons as blue might be better.

Thank you both for your suggestions. 😀

**Reply by Darthagnon** [reply]
I'll upload a v3 zip of coloured icons (hopefully fix the problems and colours with my previous recoloured versions). and a red Cancel. I'll see if I can fix the DB platter, too.

Something else I thought of - the ```Backup``` button is the one with the down arrow, and the ```Restore``` button is the one with the up arrow. Would it make more sense if Back**up** had the up arrow? (I'm personally uncertain - if it would, I can recolour, and this time record the hue changes I make, for consistency).

**Reply by MikeMaximus** [reply]
I feel like down arrow for backup makes more sense, as it's creating a file inside a folder, and a restore is pulling from it.  I was also thinking that the Save button might look better if it was a deeper purple? 

Thanks for helping out with icons!

**Reply by Darthagnon** [reply]
Glad to be of service! Thank you for making GBM; I'd be lost without it!
Is something like 8c53a1, as used by Paint.NET's save icon good?


![PaintDotNet_save icon example](https://user-images.githubusercontent.com/20792748/105508819-bda53b80-5cc4-11eb-8c9a-fd136dad7139.png)

**Reply by MikeMaximus** [reply]
Sure that looks fine

**Reply by MikeMaximus** [reply]
Any progress on the v3 icons @Darthagnon?

I don't want to seem pushy, I'm just currently updating the manual for v125 and would like to use screenshots that include the final recoloured icons.

**Reply by Darthagnon** [reply]
Please forgive me for the delay; I've been having trouble accessing my computer. I'll upload them here tomorrow, 29/01/2021.

**Reply by Darthagnon** [reply]
WIP post (I'll upload the full zip here once done, as well as icons as I do them, as I'm not sure when I'll be called away from my PC):

Darker save icon (Amazing how much difference an IPS display makes - on the other PC I was using, the garish pink of v2 looked normal):
frmMain_working
![frmMain_Working_v3](https://user-images.githubusercontent.com/20792748/106309689-653ce380-625a-11eb-8159-06a02b7029e0.png) CRC32: f92bc34a, 1.77kb

24px supersampling (paint.net "max quality") resize of the above for multisave icon:
Multi_save
![Multi_Save_v3](https://user-images.githubusercontent.com/20792748/106310038-e4cab280-625a-11eb-8054-51848b29408a.png) CRC32: be589596, 866 bytes

"Link" icon - outlined, so it stands out more (took ages, had to deconstruct the original):
frmGameManager_Link
![frmGameManager_Link_v3](https://user-images.githubusercontent.com/20792748/106315021-83a6dd00-6262-11eb-8ecc-888987f5b6d5.png) CRC32: 6a135032, 1.22kb

Cancel icons:
frmMain_Cancel_Small
![frmMain_Cancel_Small_v3](https://user-images.githubusercontent.com/20792748/106315622-6aeaf700-6263-11eb-9284-5364f9864cfb.png)
Multi_Cancel
![Multi_Cancel_v3](https://user-images.githubusercontent.com/20792748/106315623-6b838d80-6263-11eb-8d48-889f846b5baa.png)

Multi_Delete
![Multi_Delete_v3](https://user-images.githubusercontent.com/20792748/106315923-d9c85000-6263-11eb-9be3-7b0061fd897c.png)

**Reply by Darthagnon** [reply]
Here's the ZIP with all the edited icons (including the ones from the previous post), and the Paint.NET PDN project files, as well as a TXT explaining the adjustments I made (might be useful for reference for future edits if more icons are added).

[v3.zip](https://github.com/MikeMaximus/gbm/files/5895919/v3.zip)

Have a nice weekend!

**Reply by MikeMaximus** [reply]
Thanks for your work on this. They look great!

**Comment by MikeMaximus** [comment]
I haven't posted any updates for a few weeks, but work continues on v125.  I've applied some the UI changes to the entire app, and the Main Form and Game Manager have evolved to look much better than the examples posted above.  I'm at the stage of tweaking, testing and fixing issues.

A mostly up to date list of changes is always available in the update branch:
https://raw.githubusercontent.com/MikeMaximus/gbm/v125/GBM/readme.txt

I also tried to port GBM to the new and shiny .NET 5, which has some features I'm interested in.  The porting process actually went better than I thought it would, but still failed in the end.  It builds, but always crashes out on the first line of code that does anything.  Due to the vagueness of the exception(s) being thrown I can't even figure out what's making it crash, which is extremely frustrating. 😡

**Reply by Darthagnon** [reply]
I've built "pre-release" binaries from the source that I'm using at the moment; while more recent builds (maybe due to .NET 5? I didn't have it installed) don't work for me, with debug errors, some older commits work stable enough for me. I'm currently using:

- **I use this one at the moment:** GBM_v1.25beta1 pre-release, commit d9de90ab31f649d9ef523412ba5ef4bd93b07baa (last version before tabs)
- GBM_v1.25beta2 pre-release, commit 88a9d795afea8eac8f75a51549d526fb6dc2e16a (also builds and works, has early implementation of tabs)
- GBM_v1.25beta3 pre-release, commit 9baf186a9bc60c80ac1e45959ff583f8a3839c41
  - Last commit that builds and works (next commit, "Added ability to detect games with window title", and later, doesn't work yet when I build at home)

**Reply by MikeMaximus** [reply]
Unfortunately this will happen when following an in-development branch because of how GBM applies database upgrades between versions.  It's crashing because the database is missing a new table column that was added in that commit, that column needs to be added manually to any database already flagged as v125.  

This is a risk of actively using an in-development branch, worse things could also happen. 😨 I recommend making extra backups of your gbm.s3db files as well before trying any in-development version.

Easy Fix - **You'll lose the changes made since using an in-development build**

1.  Restore the automatically labelled v124 backups ```(gbm.s3db.v124.bak)```, in both your Backup folder and in ```%LOCALAPPDATA%\gbm```.
2.  Build and run the latest commit, GBM will make new backups when it re-runs the upgrade.
3.  This has the positive side effect of helping me test the database upgrade routine during development.

Note - This won't affect existing backup files, but GBM may no longer list them or give warnings because checksums no longer match the database.

Difficult Fix - **You won't lose any changes, but requires knowledge of SQL and tools**

1.  Look at the diff of mgrSQLite.vb in the commit that started crashing to see what database changes were made.
2.  Manually apply the database changes from mgrSQLite.vb using a tool such as [DB Browser for SQLite](https://sqlitebrowser.org/).
3.  This might not always be straightforward, but it usually is.

v125 is still using .NET 4.0, I didn't push any commits of the .NET 5 disaster.
